### PR TITLE
0.2.3

### PR DESCRIPTION
**Pull request by ruhrie**

### CODE REVIEW

File: TraceInfrastructure/Backend/CMakeLists.txt [hunk 1]
```diff
@@ -9,6 +9,8 @@ set_target_properties(
     OUTPUT_NAME AtlasBackend
 )
 
+target_link_options(AtlasBackend INTERFACE "-fuse-ld=lld-9")
```
Comment (benroywillis) [inline TraceInfrastructure/Backend/CMakeLists.txt:12]: If you want to use this you need to change the minimum CMake version to 3.13

**Comment by wohlbier** [comment]
I notice that you put `-9` on your `clang` and `llvm` commands. I manage my development environment with `spack` https://github.com/spack/spack. The installed binaries do not have these extensions (see below). Not sure how generic you are trying to be, but I have to go in and remove them.

```
(base) jgwohlbier@sdh01:~/.../TraceAtlas$ module list
Currently Loaded Modulefiles:
 1) emacs-26.1-gcc-7.3.0-ecuxiqe       4) ncurses-6.2-clang-9.0.1-uyyctmv  
 2) cmake-3.17.1-clang-9.0.1-juseoii   5) papi-6.0.0.1-gcc-10.1.0-adtmvr5  
 3) zlib-1.2.11-clang-9.0.1-cnj5zyp    6) llvm-9.0.1-clang-9.0.1-mupweti   
(base) jgwohlbier@sdh01:~/.../TraceAtlas$ which clang
/DATA/SDH/packages/spack/opt/spack/linux-ubuntu18.04-broadwell/clang-9.0.1/llvm-9.0.1-mupwetisd3upwdfojfn6ztdmxmgfy3kz/bin/clang
(base) jgwohlbier@sdh01:~/.../TraceAtlas$ ls /DATA/SDH/packages/spack/opt/spack/linux-ubuntu18.04-broadwell/clang-9.0.1/llvm-9.0.1-mupwetisd3upwdfojfn6ztdmxmgfy3kz/bin
bugpoint                  lldb-instr       llvm-mca
c-index-test              lldb-mi          llvm-modextract
clang                     lldb-server      llvm-mt
clang++                   lldb-vscode      llvm-nm
clang-9                   lld-link         llvm-objcopy
clang-apply-replacements  lli              llvm-objdump
clang-change-namespace    llvm-addr2line   llvm-opt-report
clang-check               llvm-ar          llvm-pdbutil
clang-cl                  llvm-as          llvm-profdata
clang-cpp                 llvm-bcanalyzer  llvm-ranlib
clangd                    llvm-cat         llvm-rc
clang-doc                 llvm-cfi-verify  llvm-readelf
clang-extdef-mapping      llvm-config      llvm-readobj
clang-format              llvm-cov         llvm-rtdyld
clang-import-test         llvm-c-test      llvm-size
clang-include-fixer       llvm-cvtres      llvm-split
clang-offload-bundler     llvm-cxxdump     llvm-stress
clang-query               llvm-cxxfilt     llvm-strings
clang-refactor            llvm-cxxmap      llvm-strip
clang-rename              llvm-diff        llvm-symbolizer
clang-reorder-fields      llvm-dis         llvm-tblgen
clang-scan-deps           llvm-dlltool     llvm-undname
clang-tidy                llvm-dwarfdump   llvm-xray
diagtool                  llvm-dwp         modularize
dsymutil                  llvm-elfabi      obj2yaml
find-all-symbols          llvm-exegesis    opt
git-clang-format          llvm-extract     sancov
hmaptool                  llvm-jitlink     sanstats
ld64.lld                  llvm-lib         scan-build
ld.lld                    llvm-link        scan-view
llc                       llvm-lipo        verify-uselistorder
lld                       llvm-lto         wasm-ld
lldb                      llvm-lto2        yaml2obj
lldb-argdumper            llvm-mc
```

**Comment by ruhrie** [comment]
I use the -9 because the versions from the llvm apt servers have them (https://apt.llvm.org/). There is no reason they can't be removed though. As long as we are using at least version 9 we are fine.

I'll investigate what needs patching.

**Comment by wohlbier** [comment]
> I use the -9 because the versions from the llvm apt servers have them (https://apt.llvm.org/). There is no reason they can't be removed though. As long as we are using at least version 9 we are fine.
> 
> I'll investigate what needs patching.

I suppose a variable could be defined, e.g., `CLANG_TIDY=clang-tidy-9` or `CLANG_TIDY=clang-tidy`?

**Comment by ruhrie** [comment]
I intend to use the cmake variables introduced by the LLVM and Clang cmake export files. They define the path to these tools and various flags that may be necessary. I just need to test them first, and they aren't well documented. They appear to introduce the executables and libraries as targets. That way I can specify to use clang with version 9 and use lld version 9 directly.

An alternative is find_program which I currently use for clang-tidy, but that has the problem that I can't guarantee version. May not really be an issue in that I can force it to be the responsibility of the user to make sure the tools are right.

I'll be testing them today/this weekend and hopefully have a good solution.

**Comment by ruhrie** [comment]
I've just made a change that I think fixed your problem. Its in master now. Feel free to give it a try.

**Comment by wohlbier** [comment]
Your method didn't work with a Spack `clang` where the CMake config file was located in `/path/to/spack/opt/spack/linux-ubuntu18.04-broadwell/clang-9.0.1/llvm-9.0.1-mupwetisd3upwdfojfn6ztdmxmgfy3kz/lib/cmake/llvm/LLVMConfig.cmake`.

Instead of `PARENT_DIR` I used `LLVM_INSTALL_PREFIX` that was picked up from the find script and that works for me. Diff below.

```
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 6aeff84..5e33a60 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -12,18 +12,17 @@ llvm_map_components_to_libnames(llvm_libs support core irreader bitreader bitwri
 #use the llvm toolchain
 #we fix these ourselves since we don't support gcc anyway
 #I want windows support, but passes on windows are tricky anyway so that is a secondary goal
-get_filename_component(PARENT_DIR ${LLVM_DIR} DIRECTORY)
-set(CMAKE_C_COMPILER ${PARENT_DIR}/bin/clang)
-set(CMAKE_CXX_COMPILER ${PARENT_DIR}/bin/clang++)
-set(CMAKE_LINKER ${PARENT_DIR}/bin/ld.lld)
-set(CMAKE_AR ${PARENT_DIR}/bin/llvm-ar)
-set(CMAKE_ADDR2LINE ${PARENT_DIR}/bin/llvm-addr2line)
-set(CMAKE_NM ${PARENT_DIR}/bin/llvm-nm)
-set(CMAKE_OBJCOPY ${PARENT_DIR}/bin/llvm-objcopy)
-set(CMAKE_OBJDUMP ${PARENT_DIR}/bin/llvm-objdump)
-set(CMAKE_RANLIB ${PARENT_DIR}/bin/llvm-ranlib)
-set(CMAKE_READELF ${PARENT_DIR}/bin/llvm-readelf)
-set(CMAKE_STRIP ${PARENT_DIR}/bin/llvm-strip)
+set(CMAKE_C_COMPILER ${LLVM_INSTALL_PREFIX}/bin/clang)
+set(CMAKE_CXX_COMPILER ${LLVM_INSTALL_PREFIX}/bin/clang++)
+set(CMAKE_LINKER ${LLVM_INSTALL_PREFIX}/bin/ld.lld)
+set(CMAKE_AR ${LLVM_INSTALL_PREFIX}/bin/llvm-ar)
+set(CMAKE_ADDR2LINE ${LLVM_INSTALL_PREFIX}/bin/llvm-addr2line)
+set(CMAKE_NM ${LLVM_INSTALL_PREFIX}/bin/llvm-nm)
+set(CMAKE_OBJCOPY ${LLVM_INSTALL_PREFIX}/bin/llvm-objcopy)
+set(CMAKE_OBJDUMP ${LLVM_INSTALL_PREFIX}/bin/llvm-objdump)
+set(CMAKE_RANLIB ${LLVM_INSTALL_PREFIX}/bin/llvm-ranlib)
+set(CMAKE_READELF ${LLVM_INSTALL_PREFIX}/bin/llvm-readelf)
+set(CMAKE_STRIP ${LLVM_INSTALL_PREFIX}/bin/llvm-strip)
 
 #other dependencies, provided by vcpkg but you can get them elsewhere as long as you have a cmake file
 find_package(ZLIB REQUIRED)
```

**Comment by ruhrie** [comment]
Looks good to me. Forgot about than variable and my method worked on all my test platforms. I honestly hadn't heard about spank before, but I'll look into it for the future.

This should reach master today.

**Pull request merged**